hazelcast-client/src/main/java/com/hazelcast/client/spi/ClientProxy.java

```diff
@@ -143,31 +143,39 @@ protected String getDistributedObjectName() {
 
     @Override
     public final void destroy() {
-        if (preDestroy()) {
-            onDestroy();
-            ClientMessage clientMessage = ClientDestroyProxyCodec.encodeRequest(getDistributedObjectName(), getServiceName());
-            getContext().getProxyManager().removeProxy(getServiceName(), getDistributedObjectName());
-            try {
-                new ClientInvocation(getClient(), clientMessage, getName()).invoke().get();
-                postDestroy();
-            } catch (Exception e) {
-                throw rethrow(e);
-            }
-        }
+        getContext().getProxyManager().destroyProxy(this);
     }
 
     /**
-     * Destroys this client proxy instance locally without issuing distributed
-     * object destroy request to the cluster as the {@link #destroy} method
-     * does.
+     * Destroys this proxy.
      * <p>
-     * The local destruction operation still may perform some communication
-     * with the cluster; for example, to unregister remote event subscriptions.
+     * Optionally, the remote distributed object counterpart also can be
+     * destroyed by passing {@code true} for the {@code alsoRemotely} parameter.
+     * In this case, a distributed object destruction operation will be issued
+     * to the cluster.
+     * <p>
+     * However, even a pure local-only destruction still may involve some
+     * communication with the cluster; for example, to unregister remote event
+     * subscriptions.
+     *
+     * @param alsoRemotely {@code false} to perform the local destruction only,
+     *                     {@code true} to destroy the remote distributed
+     *                     counterpart in addition to the local destruction.
+     * @see ProxyManager#destroyProxyLocally(String, String)
+     * @see ProxyManager#destroyProxy(ClientProxy)
      */
-    public final void destroyLocally() {
+    public final void destroy(boolean alsoRemotely) {
```

Comment:
I'm a bit confused with the naming. There are two `destroy(..)` methods; one from coming interface without arguments and the other one is this. 

As I see, `destroy(true)` is called with `true` argument only from `ProxyManager. destroyProxy(..)`. Can we move remote proxy destruction to `ProxyManager` as an internal method and rename this method to  `destroyLocally()` without any remote interaction?

In this case, `ProxyManager. destroyProxy(..)` will be like;

```java
if (registeredProxyFuture != null) {
    ClientProxy registeredProxy = registeredProxyFuture.get();
	// with some exception handling...
  	registeredProxy.destroyLocally();
   	destroyRemotely(registeredProxy); // new internal remote destruction method
    if (proxy == registeredProxy) {
        return;
    }
}

proxy.destroyLocally();
```

Reply:
Good point. My only concern is the ordering, `preDestroy`, `onDestroy`, `destroyRemotely`, `postDestroy` will become `preDestroy`, `onDestroy`, `postDestroy`, `destroyRemotely`, but seems like that should not be a problem.